### DISCUSSION TITLE
Milo Dev1.0 is HasAddIn ReferenceType Supported?

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1268] [General]
Hi we migrated our application to milo dev1.0 and i want to add a Node to our Server with the HasAddIn Reference Type but it doesnt show up in Prosys or UA Expert.  
```
private void addHasAddInTypeNode(UaFolderNode rootFolder) {
UaObjectNode foo = UaObjectNode.build(getNodeContext(), b ->
				b.setNodeId(newNodeId("Part HKL/MyObjectFoo"))
				 .setBrowseName(newQualifiedName("Foo"))
				 .setDisplayName(LocalizedText.english("Foo"))
				 .setTypeDefinition(NodeIds.BaseDataVariableType)
				 .build()
		);
getNodeManager().addNode(foo);

foo.addReference(new Reference(
				foo.getNodeId(),
				NodeIds.HasAddIn,
				rootFolder.getNodeId().expanded(),
				false
		));
}
```
Do i miss something or is it not implemented jet ? Thanks in Advance

**Comment by kevinherron** [comment]
Hmm, it probably fails somewhere in here: https://github.com/eclipse/milo/blob/a79fe4613e346db08c2f3bef96f34c6b00e10420/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/servicesets/impl/helpers/BrowseHelper.java#L397-L407

Because it hasn't been added to the [BuiltinReferenceType](https://github.com/eclipse/milo/blob/master/opc-ua-stack/stack-core/src/main/java/org/eclipse/milo/opcua/stack/core/BuiltinReferenceType.java) enum.

This subtype checking logic needs to be thrown out and replaced with a ReferenceTypeTree managed by the OpcUaServer that gets built/updated when new reference types are registered. I think started working towards this (ReferenceTypeTree and ReferenceTypeTreeBuilder exist) but haven't implemented it yet.

**Comment by johannwesely** [comment]
Hi, first i would like to thank you for your work as the dev1.0 works for us now.  We had to do the javax to jakarta migration as well to get it running. One point i stumbled over is  that the BsdGenerator in Line 29 wants to set a property to the marshaller that doesnt exist in Jakarta "com.sun.xml.bind.namespacePrefixMapper" so i get an exception for that. But Client and Server are up and running. Do you have a Timeline for me when the HasAddIn ReferenceType will be implemented in dev1.0? Greetings Johann

**Reply by kevinherron** [reply]
I'll look at this a little bit today and see if I can fix it. It may be an easy fix because I've already done the ReferenceTypeTree work.

**Comment by kevinherron** [comment]
Are you able to test this PR/branch? 

https://github.com/eclipse/milo/pull/1270
https://github.com/eclipse/milo/tree/issues/gh1269

**Comment by johannwesely** [comment]
Hi it worked i have added a Screenshot from prosys. 
![prosysscreen](https://github.com/eclipse/milo/assets/108257325/a7c3e65b-c84f-4143-b129-90aa7903592f)

Thanks you very much now ich can progess with my Tests.
I have one last question how do i add a new referenceType to the tree ? getServer().getReferenceTypes().put(node.getNodeId(), new ReferenceType(){...} becomes ? Thank you Johann

**Reply by kevinherron** [reply]
Not sure yet, haven't gotten that far into it :)

I think all that will be necessary is to make sure it's available in the AddressSpace (as an actual UaReferenceTypeNode in a UaNodeManager that was registered with the server) so that the `Lazy` holding the `ReferenceTypeTree` can be reset and then when it gets re-built it will find the new ReferenceType.

**Reply by kevinherron** [reply]
@johannwesely added another commit that has an `OpcUaServer::updateReferenceTypeTree()` method. If you add new `UaReferenceTypeNode`s after the first time anything calls `getReferenceTypeTree()` then you can call update and the tree will get re-built.
